discovery_engine_core/core/src/stack/exploration.rs

```diff
@@ -0,0 +1,531 @@
+    indices.sort_unstable_by(|&i, &j| nan_safe_f32_cmp(&arr[i], &arr[j]));
+    indices
+}
+
+/// Computes the cosine similarity between the cois and documents and returns the
+/// cosine similarity of the nearest coi for each document.
+///
+/// # Panics
+/// Panics if `cois` is empty.
+fn find_nearest_coi_for_docs(
+    docs: &[ArrayView1<'_, f32>],
+    cois: &[ArrayView1<'_, f32>],
+) -> Vec<f32> {
+    // creates a cosine similarity matrix (docs x cois)
+    // |      | coi1                  | coi2                  |
+    // | doc1 | `cos_sim(doc1, coi1)` | `cos_sim(doc1, coi2)` |
+    // | doc2 | `cos_sim(doc2, coi1)` | `cos_sim(doc2, coi2)` |
+    //
+    // finds the nearest coi for each document
+    // [doc1(max(cos_sim1, cos_sim2, ...)), doc2(max(cos_sim1, cos_sim2, ...)), ...]
+    docs.iter()
+        .cartesian_product(cois)
+        .map(|(a, b)| cosine_similarity(*a, *b))
+        .chunks(cois.len())
+        .into_iter()
+        .map(|similarities_for_doc| {
+            similarities_for_doc
+                .max_by(nan_safe_f32_cmp)
+                .unwrap_or_default()
+        })
+        .collect()
+}
+
+/// Determines the threshold and returns it together with the indices of the documents
+/// that are below that threshold.
+///
+/// If `number_of_candidates > nearest_coi_for_docs.len()`, `number_of_candidates`
+/// will be the value of `nearest_coi_for_docs.len()`.
+///
+/// If `number_of_candidates` or `nearest_coi_for_docs.len()` equals `0`, the function will
+/// return a threshold of `0.` and an empty set of candidates.
+fn select_initial_candidates(
+    nearest_coi_for_docs: &[f32],
+    number_of_candidates: usize,
+) -> (f32, BTreeSet<usize>) {
+    let number_of_candidates = number_of_candidates.min(nearest_coi_for_docs.len());
+    if number_of_candidates == 0 {
+        return (0., BTreeSet::new());
+    }
+
+    let mut indices = argsort(nearest_coi_for_docs);
+    let threshold = indices
+        .get(number_of_candidates - 1 /* safe because we check before that number_of_candidates != 0 */)
+        .map(|idx| nearest_coi_for_docs.get(*idx).unwrap(/* safe because it's an index from the argsort of nearest_coi_for_docs */))
+        .unwrap(/* safe because number_of_candidates <= nearest_coi_for_docs.len() */);
+    let candidates = indices.drain(..number_of_candidates).collect();
+    (*threshold, candidates)
+}
+
+/// Retains the documents that have been selected.
+fn retain_documents_by_indices(
+    mut selected: Vec<usize>,
+    documents: Vec<Document>,
+) -> Vec<Document> {
+    selected.sort_unstable_by_key(|w| Reverse(*w));
+    let mut retain = Vec::with_capacity(selected.len());
+    for (idx, doc) in documents.into_iter().enumerate() {
+        if selected.is_empty() {
+            break;
+        } else if idx == *selected.last().unwrap(/* can't be empty because of the check above */) {
+            retain.push(doc);
+            selected.pop();
+        }
+    }
+    retain
```

Comment:
:art: not sure if it makes a difference perfomance-wise, but imo the whole sort-push-pop-loop is difficult to read, wdyt about the following instead? the order of the retained documents will be different though.
```suggestion
    documents
        .into_iter()
        .enumerate()
        .filter_map(|(idx, doc)| selected.contains(&idx).then(|| doc))
        .collect()
```

Reply:
done!